### PR TITLE
Cache loopy kernels for faster look-up

### PR DESCRIPTION
**Pull request by tkarna**
Loopy kernels are slower to evaluate than C kernels mostly due to the overhead in `loopy.make_function` pipeline. The fix (proposed by @wence-) is to cache the loopy kernels to reduce the overhead.

### CODE REVIEW

**Comment by tkarna** [comment]
For the record, here's a test comparing slope limiter C and loopy kernels. The C kernel runs approximately 50x faster (tested with firedrake commit 4455fe10).

```python
from firedrake import *
import timeit

n = 10
mesh = UnitSquareMesh(n, n)

P1DG = FunctionSpace(mesh, 'DG', 1)
P0 = FunctionSpace(mesh, 'DG', 0)
P1CG = FunctionSpace(mesh, 'CG', 1)

centroids = Function(P0)
centroids_rhs = Function(P0)
max_field = Function(P1CG)
min_field = Function(P1CG)

field = Function(P1DG, name='f').assign(3.3)

domain = "{[i, ii]: 0 <= i < q.dofs and 0 <= ii < q.dofs}"
instructions = """
<float64> alpha = 1
<float64> qavg = qbar[0, 0]
for i
    <float64> _alpha1 = fmin(alpha, fmin(1, (qmax[i] - qavg)/(q[i] - qavg)))
    <float64> _alpha2 = fmin(alpha, fmin(1, (qavg - qmin[i])/(qavg - q[i])))
    alpha = _alpha1 if q[i] > qavg else (_alpha2 if q[i] < qavg else  alpha)
end
for ii
    q[ii] = qavg + alpha * (q[ii] - qavg)
end
"""
loopy_kernel = (domain, instructions)


c_kernel = """
double alpha = 1.0;
double qavg = qbar[0];
for (int i=0; i < q.dofs; i++) {
    if (q[i] > qavg)
        alpha = fmin(alpha, fmin(1, (qmax[i] - qavg)/(q[i] - qavg)));
    else if (q[i] < qavg)
        alpha = fmin(alpha, fmin(1, (qavg - qmin[i])/(qavg - q[i])));
}
for (int i=0; i<q.dofs; i++) {
    q[i] = qavg + alpha*(q[i] - qavg);
}
"""


def exe_loopy():
    par_loop(
        loopy_kernel, dx,
        {"qbar": (centroids, READ),
         "q": (field, RW),
         "qmax": (max_field, READ),
         "qmin": (min_field, READ)
        },
        is_loopy_kernel=True
    )


def exe_c():
    par_loop(
        c_kernel, dx,
        {"qbar": (centroids, READ),
            "q": (field, RW),
            "qmax": (max_field, READ),
            "qmin": (min_field, READ)
        },
        is_loopy_kernel=False
    )

n = 100
t_loopy = timeit.timeit('exe_loopy()', setup='from __main__ import exe_loopy', number=n)
t_c = timeit.timeit('exe_c()', setup='from __main__ import exe_c', number=n)

print(f'loopy {t_loopy:.3f}')
print(f'c {t_c:.3f}')
print(f'ratio {t_loopy/t_c:.3f}')
```

File: firedrake/parloops.py [hunk 1]
```diff
@@ -103,11 +109,20 @@ def _form_loopy_kernel(kernel_domains, instructions, measure, args, **kwargs):
...
+        else:
+            raise KeyError("No cache")
+    except KeyError:
+        kargs.append(...)
+        knl = loopy.make_function(kernel_domains, instructions, kargs, seq_dependencies=True,
+                                  name="par_loop_kernel", silenced_warnings=["summing_if_branches_ops"], target=loopy.CTarget())
+        if kernel_cache is not None:
+            return kernel_cache.setdefault(key, pyop2.Kernel(knl, "par_loop_kernel", **kwargs))
+        else:
+            return knl
```
Comment (wence-) [inline firedrake/parloops.py:125]: I screwed this bit up. I think it want's to be
```suggestion
        knl = pyop2.Kernel(knl, "par_loop_kernel", **kwargs)
        if kernel_cache is not None:
            return kernel_cache.setdefault(key, knl)
        else:
            return knl
```

Comment (tkarna) [reply firedrake/parloops.py]: oh yes. fixed now.

**Comment by wence-** [comment]
Thanks

**Pull request merged**